File: library/src/main/java/com/lee/fileselector/FileAdapter.java
package com.lee.fileselector;

import android.app.Activity;
import android.content.Context;
import android.view.MenuItem;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.lee.adapter.recyclerview.CommonAdapter;
import com.lee.adapter.recyclerview.base.ViewHolder;
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Lee
 */
public class FileAdapter extends CommonAdapter<FileModel> implements FastScrollRecyclerView.SectionedAdapter {

    private MenuItem mCountMenuItem;

    private ArrayList<FileModel> mSelectedFileList;

    private List<FileModel> modelList;

    public void setCountMenuItem(MenuItem mCountMenuItem) {
        this.mCountMenuItem = mCountMenuItem;
    }

    public void setSelectedFileList(ArrayList<FileModel> mSelectedFileList) {
        this.mSelectedFileList = mSelectedFileList;
    }

    public FileAdapter(Context context, int layoutId, List<FileModel> datas) {
        super(context, layoutId, datas);
        modelList = datas;
    }

    @Override
    public void convert(ViewHolder holder, FileModel fileModel, int position) {
        ImageView imageView = holder.getView(R.id.iv_type);
        String extension = fileModel.getExtension();
        if (extension.equalsIgnoreCase("gif")) {
            Glide.with(mContext).asGif().load(fileModel.getPath()).into(imageView);
        } else if (extension.equalsIgnoreCase("jpg") ||
                extension.equalsIgnoreCase("png") ||
                extension.equalsIgnoreCase("bmp") ||
                extension.equalsIgnoreCase("jpeg") ||
                extension.equalsIgnoreCase("webp") ||
                extension.equalsIgnoreCase("wmv") ||
                extension.equalsIgnoreCase("flv") ||
                extension.equalsIgnoreCase("mp4") ||
                extension.equalsIgnoreCase("avi") ||
                extension.equalsIgnoreCase("mpg") ||
                extension.equalsIgnoreCase("rmvb") ||
                extension.equalsIgnoreCase("rm") ||
                extension.equalsIgnoreCase("asf") ||
                extension.equalsIgnoreCase("f4v") ||
                extension.equalsIgnoreCase("vob") ||
                extension.equalsIgnoreCase("mkv") ||
                extension.equalsIgnoreCase("3gp") ||
                extension.equalsIgnoreCase("mov")) {
            Glide.with(mContext).load(fileModel.getPath()).into(imageView);
        } else {
            String s = AvatarUtils.generateDefaultAvatar(mContext, extension);
            Glide.with(mContext).load(s).into(imageView);
        }

        SmoothCheckBox checkBox = holder.getView(R.id.checkbox);
        RelativeLayout layout = holder.getView(R.id.layout_item);
        holder.setText(R.id.tv_name, fileModel.getName());
        holder.setText(R.id.tv_detail, FileUtils.getDateTime(fileModel.getDate()) + "  -  " + formatFileSize(fileModel.getSize()));
        checkBox.setOnCheckedChangeListener(null);
        checkBox.setChecked(fileModel.isSelected(), false);
        checkBox.setOnCheckedChangeListener(new SmoothCheckBox.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(SmoothCheckBox checkBox, boolean isChecked) {
                if (!isChecked && fileModel.isSelected()) {
                    int index = findFileIndex(fileModel);
                    if (index != -1) {
                        mSelectedFileList.remove(index);
                    }
                    fileModel.setSelected(false);
                } else if (isChecked && !fileModel.isSelected()) {
                    if (mSelectedFileList.size() >= FileSelector.getInstance((Activity) mContext).maxCount) {
                        Toast.makeText(mContext, "您最多只能选择" + FileSelector.getInstance((Activity) mContext).maxCount + "个", Toast.LENGTH_SHORT).show();
                        checkBox.setChecked(false, true);
                        return;
                    }
                    mSelectedFileList.add(fileModel);
                    fileModel.setSelected(true);
                }
                mCountMenuItem.setTitle(String.format(mContext.getString(R.string.selected_file_count), String.valueOf(mSelectedFileList.size()), String.valueOf(FileSelector.getInstance((Activity) mContext).maxCount)));
            }
        });
        layout.setOnClickListener(v -> {
            if (fileModel.isSelected()) {
                int index = findFileIndex(fileModel);
                if (index != -1) {
                    mSelectedFileList.remove(index);
                }
                fileModel.setSelected(false);
            } else {
                if (mSelectedFileList.size() >= FileSelector.getInstance((Activity) mContext).maxCount) {
                    Toast.makeText(mContext, "您最多只能选择" + FileSelector.getInstance((Activity) mContext).maxCount + "个", Toast.LENGTH_SHORT).show();
                    return;
                }
                mSelectedFileList.add(fileModel);
                fileModel.setSelected(true);
            }
            checkBox.setChecked(fileModel.isSelected(), true);
            mCountMenuItem.setTitle(String.format(mContext.getString(R.string.selected_file_count), String.valueOf(mSelectedFileList.size()), String.valueOf(FileSelector.getInstance((Activity) mContext).maxCount)));
        });
    }

    @NonNull
    @Override
    public String getSectionName(int position) {
        String namePinyin = CharacterParser.getInstance().getSpelling(modelList.get(position).getName());
        return namePinyin.substring(0, 1).toUpperCase();
    }

    public String formatFileSize(long size) {
        DecimalFormat df = new DecimalFormat("#.00");
        String fileSizeString;
        if (size < 1024) {
            fileSizeString = df.format((double) size) + "B";
        } else if (size < 1048576) {
            fileSizeString = df.format((double) size / 1024) + "K";
        } else if (size < 1073741824) {
            fileSizeString = df.format((double) size / 1048576) + "M";
        } else {
            fileSizeString = df.format((double) size / 1073741824) + "G";
        }
        return fileSizeString;
    }

    private int findFileIndex(FileModel item) {
        for (int i = 0; i < mSelectedFileList.size(); i++) {
            if (mSelectedFileList.get(i).getPath().equals(item.getPath())) {
                return i;
            }
        }
        return -1;
    }
}

File: library/src/main/kotlin/xyz/junerver/fileselector/FileAdapter.kt
package xyz.junerver.fileselector

import android.content.Context
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.mikhaellopez.circularprogressbar.CircularProgressBar
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView
import xyz.junerver.fileselector.utils.AvatarUtils
import xyz.junerver.fileselector.utils.CharacterParserUtils
import xyz.junerver.fileselector.worker.DIND_TALK_PATH
import java.util.ArrayList

/**
 * Description:
 * @author Junerver
 * date: 2022/2/16-14:53
 * Email: junerver@gmail.com
 * Version: v1.0
 */
class FileAdapter(
    private val mContext: Context,
    private val layoutId: Int,
    private val modelList: List<FileModel>,
    //是否为选择器模式
    private val isSelectorMode: Boolean = false
) : RecyclerView.Adapter<FileAdapter.ViewHolder>(), FastScrollRecyclerView.SectionedAdapter {

    //最大选择数量
    private var mMaxSelect = FileSelector.maxCount

    private val mFileTypeBitmapMap = HashMap<String, String>()

    //主UI的右上角menu
    private var mCountMenuItem: MenuItem? = null

    //选择的文件
    private var mSelectedFileList: ArrayList<FileModel>? = null

    interface BrowserItemOnClickListener {
        fun onItemClick(ctx: Context, holder: ViewHolder, fileModel: FileModel)
        fun onItemLongClick(ctx: Context, holder: ViewHolder, fileModel: FileModel)
    }

    private var mBrowserItemOnClickListener: BrowserItemOnClickListener? = null

    fun setListener(onClickListener: BrowserItemOnClickListener) {
        mBrowserItemOnClickListener = onClickListener
    }

    fun setMaxSelect(max: Int) {
        mMaxSelect = max
    }

    fun setCountMenuItem(mCountMenuItem: MenuItem?) {
        this.mCountMenuItem = mCountMenuItem
    }

    fun setSelectedFileList(mSelectedFileList: ArrayList<FileModel>?) {
        this.mSelectedFileList = mSelectedFileList
    }

    override fun getSectionName(position: Int): String {
        val namePinyin = CharacterParserUtils.getSpelling(
            modelList[position].name
        )
        return namePinyin.substring(0, 1).uppercase()
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val root = mContext.inflater(R.layout.item_file_selector, parent, false)
        return ViewHolder(root)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val fileModel = modelList[position]
        val imageView = holder.imageView
        val extension = fileModel.extension
        if (extension.equals("gif", ignoreCase = true) ||
            extension.equals("jpg", ignoreCase = true) ||
            extension.equals("png", ignoreCase = true) ||
            extension.equals("bmp", ignoreCase = true) ||
            extension.equals("jpeg", ignoreCase = true) ||
            extension.equals("webp", ignoreCase = true) ||
            extension.equals("wmv", ignoreCase = true) ||
            extension.equals("flv", ignoreCase = true) ||
            extension.equals("mp4", ignoreCase = true) ||
            extension.equals("avi", ignoreCase = true) ||
            extension.equals("mpg", ignoreCase = true) ||
            extension.equals("rmvb", ignoreCase = true) ||
            extension.equals("rm", ignoreCase = true) ||
            extension.equals("asf", ignoreCase = true) ||
            extension.equals("f4v", ignoreCase = true) ||
            extension.equals("vob", ignoreCase = true) ||
            extension.equals("mkv", ignoreCase = true) ||
            extension.equals("3gp", ignoreCase = true) ||
            extension.equals("mov", ignoreCase = true)
        ) {
            if (fileModel.isAndroidData) {
                imageView.load(fileModel.uri)
            } else {
                imageView.load(fileModel.path)
            }
        } else {
            val s = if (mFileTypeBitmapMap.containsKey(extension)) {
                mFileTypeBitmapMap[extension]
            } else {
                val bitmapPath = AvatarUtils.generateDefaultAvatar(mContext, extension)
                mFileTypeBitmapMap[extension] = bitmapPath
                bitmapPath
            }
            Glide.with(mContext).load(s).into(imageView)
        }
        holder.apply {
            name.text = fileModel.name
            detail.text = "${getDateTime(fileModel.date)}  -  ${formatFileSize(fileModel.size)}"
        }
        //勾选框配置
        val checkBox = holder.checkBox
        checkBox.apply {
            setOnCheckedChangeListener(null)
            setChecked(fileModel.isSelected, false)
            setOnCheckedChangeListener(SmoothCheckBox.OnCheckedChangeListener { cb, isChecked ->
                if (!isChecked && fileModel.isSelected) {
                    mSelectedFileList?.remove(fileModel)
                    fileModel.isSelected = false
                } else if (isChecked && !fileModel.isSelected) {
                    if (mSelectedFileList?.size?.let { it >= mMaxSelect } == true) {
                        Toast.makeText(
                            mContext,
                            "您最多只能选择" + FileSelector.maxCount.toString() + "个",
                            Toast.LENGTH_SHORT
                        ).show()
                        cb.setChecked(false, true)
                        return@OnCheckedChangeListener
                    }
                    "onCheckedChanged: ${fileModel.name}".log()
                    mSelectedFileList?.also {
                        it += fileModel
                    }
                    fileModel.isSelected = true
                }
                mCountMenuItem?.title = String.format(
                    mContext.getString(R.string.selected_file_count),
                    mSelectedFileList?.size ?: "0",
                    java.lang.String.valueOf(FileSelector.maxCount)
                )
            })
            if (isSelectorMode) {
                visible()
            } else {
                gone()
            }
        }
        //data目录标志
        val flag = holder.flag
        flag.apply {
            if (fileModel.isAndroidData) {
                FileSelector.ICON_MAP[getFileParentPath(fileModel.path)]?.let {
                    setImageDrawable(this.context.getDrawableRes(it))
                    this.visible()
                } ?: run { gone() }
            } else {
                if (fileModel.path.contains(DIND_TALK_PATH)) {
                    setImageDrawable(this.context.getDrawableRes(R.drawable.icon_dingding))
                    this.visible()
                } else {
                    gone()
                }
            }
        }
        val layout = holder.layout
        layout.apply {
            setOnClickListener {
                if (isSelectorMode) {
                    if (fileModel.isSelected) {
                        mSelectedFileList?.remove(fileModel)
                        fileModel.isSelected = false
                    } else {
                        if (mSelectedFileList?.size?.let { size -> size >= mMaxSelect } == true) {
                            Toast.makeText(
                                mContext,
                                "您最多只能选择" + FileSelector.maxCount.toString() + "个",
                                Toast.LENGTH_SHORT
                            ).show()
                            return@setOnClickListener
                        }
                        mSelectedFileList?.also { list -> list += fileModel }
                        fileModel.isSelected = true
                    }
                    checkBox.setChecked(fileModel.isSelected, true)
                    mCountMenuItem?.title = String.format(
                        mContext.getString(R.string.selected_file_count),
                        mSelectedFileList?.size ?: "0",
                        java.lang.String.valueOf(FileSelector.maxCount)
                    )
                } else {
                    //打开
                    mBrowserItemOnClickListener?.onItemClick(it.context, holder, fileModel)
                }
            }

            setOnLongClickListener {
                if (!isSelectorMode) {
                    mBrowserItemOnClickListener?.onItemLongClick(it.context, holder, fileModel)
                }
                true
            }
        }
    }

    override fun getItemCount(): Int {
        return modelList.size
    }

    override fun onViewRecycled(holder: ViewHolder) {
        holder.imageView.apply {
            this.setImageDrawable(null)
            Glide.with(this).clear(this)
        }
        super.onViewRecycled(holder)
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val imageView: ImageView = itemView.findViewById(R.id.iv_type)
        val checkBox: SmoothCheckBox = itemView.findViewById(R.id.checkbox)
        val flag: ImageView = itemView.findViewById(R.id.iv_data_flag)
        val layout: RelativeLayout = itemView.findViewById(R.id.layout_item)
        val name: TextView = itemView.findViewById(R.id.tv_name)
        val detail: TextView = itemView.findViewById(R.id.tv_detail)
        val circularProgressBar: CircularProgressBar =
            itemView.findViewById(R.id.circularProgressBar)
    }
}